gulpfile.js

```diff
@@ -1,29 +1,79 @@
+const { spawn } = require('child_process');
+
+const del = require('del');
 const forever = require('forever-monitor');
 const gulp = require('gulp');
+const gulpBabel = require('gulp-babel');
 const eslint = require('gulp-eslint');
 const mocha = require('gulp-mocha');
+const sourcemaps = require('gulp-sourcemaps');
+
+/**
+ * @summary Delete the dist/ directory
+ */
+const babelClean = () => del(['dist']);
+
+/**
+ * @summary Copy files that don't need to be compiled from src/ to dist/
+ */
+const babelCopy = () => gulp.src(['src/**', '!src/**/*.js', '!src/tests/integration/**'])
+  .pipe(gulp.dest('dist'));
+
+/**
+ * @summary Compile JavaScript files and place them in dist/. Also, generate sourcemaps
+ */
+const babelCompile = () => gulp.src(['src/**/*.js'])
+  .pipe(sourcemaps.init())
+  .pipe(gulpBabel())
+  .pipe(gulp.dest('dist'));
+
+const babel = gulp.series(babelClean, babelCopy, babelCompile);
 
 /**
  * @summary Use Eslint linting *.js file besides source files in node_modules
  */
-gulp.task('lint', () => gulp.src(['**/*.js', '!node_modules/**'])
+const lint = () => gulp.src(['src/**/*.js'])
   .pipe(eslint())
   .pipe(eslint.format())
-  .pipe(eslint.failAfterError()));
+  .pipe(eslint.failAfterError());
 
 /**
- * @summary Run unit test
+ * @summary Check Flow types
  */
-gulp.task('test', () => gulp.src(['tests/unit/*.js'])
-  .pipe(mocha({ reporter: 'spec' })));
+const typecheck = () => spawn('./node_modules/.bin/flow', ['check'], { stdio: 'inherit' });
 
 /**
- * @summary Start application using forever
+ * @summary Run unit tests
  */
-gulp.task('start', () => new forever.Monitor('app.js').start());
+const test = () => gulp.src(['dist/tests/unit/*.js'])
+  .pipe(mocha({ reporter: 'spec' }));
 
+/**
+ * @summary Start application using forever
+ */
+const start = () => new forever.Monitor('dist/app.js').start();
 
 /**
- * @summary Run test and lint task parallelly before start the apllication
+ * @summary Lint and compile, test, and start the application
+ */
+exports.run = gulp.series(gulp.parallel(lint, typecheck, babel), test, start);
+/**
+ * @summary Compile and start the application only
+ */
+exports.start = gulp.series(babel, start);
+/**
+ * @summary Compile and test the application only
+ */
+exports.test = gulp.series(babel, test);
+/**
+ * @summary Compile the code only
+ */
+exports.babel = babel;
+/**
+ * @summary Lint the code only
+ */
+exports.lint = lint;
+/**
+ * @summary Typecheck the code only
  */
-gulp.task('run', gulp.series(gulp.parallel('lint', 'test'), 'start'));
+exports.typecheck = typecheck;
```

Comment:
Good work here. Just curious, since we're building our own compiling workflow by ourselves, do we want to look into using some other popular package tools like webpack to do these tasks for us? Your current implementation looks totally great to me, I'm just not sure if there's any dependency outdated or has bugs how difficult it'll be for us to fix or find alternatives to replace them. Just my two cent, since webpack is the most popular packaging tool I've heard, I think their community could also be our sources in the future as well. What do you think?

Reply:
I agree, webpack seems to be pretty popular and we should look into how we could use it in our skeleton. We could look into this as part of [CO-1437](https://jira.sig.oregonstate.edu/browse/CO-1437). I'll add a comment there.